clang/unittests/Format/TokenAnnotatorTest.cpp

```diff
@@ -1071,6 +1071,16 @@ TEST_F(TokenAnnotatorTest, UnderstandsRequiresClausesAndConcepts) {
                     "concept C = (!Foo<T>) && Bar;");
   ASSERT_EQ(Tokens.size(), 19u) << Tokens;
   EXPECT_TOKEN(Tokens[15], tok::ampamp, TT_BinaryOperator);
+
+  Tokens = annotate("void f() & requires(C<decltype(x)>) {}");
+  ASSERT_EQ(Tokens.size(), 18u) << Tokens;
+  EXPECT_TOKEN(Tokens[4], tok::amp, TT_PointerOrReference);
+  EXPECT_TOKEN(Tokens[5], tok::kw_requires, TT_RequiresClause);
+
+  Tokens = annotate("auto f() -> int& requires(C<decltype(x)>) {}");
+  ASSERT_EQ(Tokens.size(), 20u) << Tokens;
+  EXPECT_TOKEN(Tokens[6], tok::amp, TT_PointerOrReference);
+  EXPECT_TOKEN(Tokens[7], tok::kw_requires, TT_RequiresClause);
 }
```

Comment:
For the mentioned test coverage (you are completely right), something like:
```c++
int foo(auto t) {
    if (requires (decltype(t) x) {x.foo();}) {
        return 1;
    }
    return 0;
}
```

Reply:
Thanks, but this test input never actually runs parseRequires(), but I constructed another case which does, I'll add both

Reply:
I found a failing edge case
`bool x = t && requires(Foo<decltype(t)> x) { x.foo(); };`
will annotate the requires as a TT_RequiresClause. I'll add extra logic in the lookahead to fix this

Reply:
Well, the simplest fix is to remove `case tok::r_paren:` on line 3424 which always led to parseRequiresClause. Surprisingly removing that breaks no tests, but surely @HazardyKnusperkeks you put it there for a reason? I'm not sure if such a fix is the best idea

Reply:
I _think_ the idea was that if it is an expression we would detect the type before reaching the `r_paren` and thus it has to be a clause. But 2 years have passed and I don't know what I was thinking last week. ;) Had to reread the code and try to guess the conditions.

Which shows why one should comment complicated code...

If I read the code right my question is why didn't `Foo<decktype(t)>` suffice `FoundType && !LastWasColonColon && OpenAngles == 0` when hitting the `x`?

On a different note one could extend the look behind to see if one finds a `=` or a `(` for any conditional, or a requires expression `}`, then it should be an expression.

Reply:
> If I read the code right my question is why didn't `Foo<decktype(t)>` suffice `FoundType && !LastWasColonColon && OpenAngles == 0` when hitting the `x`?

I'm not sure what you mean. If you're talking about `Foo<decltype(t)> x`, it never gets to the `x` because it reaches the `)` of the decltype first and that unconditionally leads to being parsed as a clause

Reply:
The fix I would originally do was just look behind again and see if the ) belongs to a decltype, I just also tried removing the case entirely, and was surprised it works. If you prefer, I can follow my original plan and add an extra check for decltype

Reply:
I have no problem with removing it. If you keep it, I would add a check for `OpenAngles == 0`. The same applies to `||` because one could write something like
```c++
requires (T<C1 || C2> t) {}
```

So the whole logic should only apply to the out of template area, similar to `const`, `volatile`, and `comma` above.

Reply:
> The fix I would originally do was just look behind again and see if the ) belongs to a decltype, I just also tried removing the case entirely, and was surprised it works. If you prefer, I can follow my original plan and add an extra check for decltype

In general, I prefer to not include code that can't be reached by test cases as the code is likely superfluous. Because @HazardyKnusperkeks is ok with its removal, wouldn't be better to just do that?